internal/image/resource_image.go

```diff
@@ -444,6 +462,168 @@ func (r ImageResource) SyncState(ctx context.Context, tfState *tfsdk.State, serv
+		return
+	}
+
+	var dataPath, metadataPath string
+	if sourceFileModel.MetadataPath.IsNull() {
+		// Unified image
+		metadataPath = sourceFileModel.DataPath.ValueString()
+	} else {
+		dataPath = sourceFileModel.DataPath.ValueString()
+		metadataPath = sourceFileModel.MetadataPath.ValueString()
+	}
+
+	var image api.ImagesPost
+	var createArgs *incus.ImageCreateArgs
+
+	imageType := "container"
+	// TODO: should https urls also be supported?
+	if strings.HasPrefix(dataPath, "https://") {
+		image.Source = &api.ImagesPostSource{}
+		image.Source.Type = "url"
+		image.Source.Mode = "pull"
+		image.Source.Protocol = "direct"
+		image.Source.URL = dataPath
+		createArgs = nil
+	} else {
+		var meta io.ReadCloser
+		var rootfs io.ReadCloser
+
+		// Open meta
+		if utils.IsDir(metadataPath) {
+			metadataPath, err = packImageDir(metadataPath)
+			if err != nil {
+				resp.Diagnostics.AddError(fmt.Sprintf("Failed to pack metadata_path: %s", metadataPath), err.Error())
+				return
+			}
+			// remove temp file
+			defer func() { _ = os.Remove(metadataPath) }()
+		}
+
+		meta, err = os.Open(metadataPath)
+		if err != nil {
+			resp.Diagnostics.AddError(fmt.Sprintf("Failed to open metadata_path: %s", metadataPath), err.Error())
+			return
+		}
+
+		defer func() { _ = meta.Close() }()
+
+		// Open rootfs
+		if dataPath != "" {
+			rootfs, err = os.Open(dataPath)
+			if err != nil {
+				resp.Diagnostics.AddError(fmt.Sprintf("failed to open data_path: %s", dataPath), err.Error())
+				return
+			}
+
+			defer func() { _ = rootfs.Close() }()
+
+			_, ext, _, err := archive.DetectCompressionFile(rootfs)
+			if err != nil {
+				resp.Diagnostics.AddError(fmt.Sprintf("Failed to detect compression of rootfs in data_path: %s", dataPath), err.Error())
+				return
+			}
+
+			_, err = rootfs.(*os.File).Seek(0, io.SeekStart)
+			if err != nil {
+				resp.Diagnostics.AddError(fmt.Sprintf("Failed to seek start for rootfas in data_path: %s", dataPath), err.Error())
+				return
+			}
+
+			if ext == ".qcow2" {
+				imageType = "virtual-machine"
+			}
+		}
+
+		createArgs = &incus.ImageCreateArgs{
+			MetaFile:   meta,
+			MetaName:   filepath.Base(metadataPath),
+			RootfsFile: rootfs,
+			RootfsName: filepath.Base(dataPath),
+			Type:       imageType,
+		}
+
+		image.Filename = createArgs.MetaName
+	}
+
+	aliases, diags := ToAliasList(ctx, plan.Aliases)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	// NOTE: this does not seem to have an effect
+	// imageAliases := make([]api.ImageAlias, 0, len(aliases))
+	// for _, alias := range aliases {
+	// 	// Ensure image alias does not already exist.
+	// 	aliasTarget, _, _ := server.GetImageAlias(alias)
+	// 	if aliasTarget != nil {
+	// 		resp.Diagnostics.AddError(fmt.Sprintf("Image alias %q already exists", alias), "")
+	// 		return
+	// 	}
+
+	// 	ia := api.ImageAlias{
+	// 		Name: alias,
+	// 	}
+
+	// 	imageAliases = append(imageAliases, ia)
+	// }
+	// image.Aliases = imageAliases
```

Comment:
I expected incus to respect the `Aliases` provided in the `api.ImagesPost` request, but this is currently not the case. Therefore I added the aliases manually, see https://github.com/lxc/terraform-provider-incus/pull/169/files#diff-dedceb2079bdc5668f6c75a9b22a13d869a760a187b2804ac22f83eeac6d50e7R611

Not sure, about the correct thing to do here.

Reply:
Pretty sure the server does, but the client code wants the extra Aliases passed through args.Aliases `ImageCopyArgs`

Reply:
What do you mean with it does not respect the aliases provided? because when I set the aliases before in a `incus_image` then I see the alias created:

```hcl
resource "incus_image" "alpine" {
  project = "default"

  aliases = ["alpine", "alpine/edge"]

  source_image = {
    remote       = "images"
    name         = "alpine/edge"
    type         = "container"
    architecture = "aarch64"
  }
}
```

```bash
$ incus image list
+-----------------+--------------+--------+------------------------------------+--------------+-----------+---------+----------------------+
|      ALIAS      | FINGERPRINT  | PUBLIC |            DESCRIPTION             | ARCHITECTURE |   TYPE    |  SIZE   |     UPLOAD DATE      |
+-----------------+--------------+--------+------------------------------------+--------------+-----------+---------+----------------------+
| alpine (1 more) | 57c20fddd750 | no     | Alpine edge arm64 (20241119_13:00) | aarch64      | CONTAINER | 3.31MiB | 2024/11/22 22:50 CET |
+-----------------+--------------+--------+------------------------------------+--------------+-----------+---------+----------------------+

$ incus image list -f json | jq .[].aliases
[
  {
    "name": "alpine",
    "description": ""
  },
  {
    "name": "alpine/edge",
    "description": ""
  }
]
```

Reply:
I tested the import from `source_file` with inucs from `main`branch (including https://github.com/lxc/incus/pull/1409) and the acceptance test is not successful, if I use the [commented out](https://github.com/lxc/terraform-provider-incus/pull/169/files#diff-dedceb2079bdc5668f6c75a9b22a13d869a760a187b2804ac22f83eeac6d50e7R571-R587) section from above without the [manual adding of the aliases](https://github.com/lxc/terraform-provider-incus/pull/169/files#diff-dedceb2079bdc5668f6c75a9b22a13d869a760a187b2804ac22f83eeac6d50e7R612-R622).

The reason in my opinion is, that the import `from_source` file does use [`client.CreateImage`](https://github.com/lxc/incus/blob/main/client/incus_images.go#L415) directly and this function does still not respect the `Aliases`. The fix from https://github.com/lxc/incus/pull/1409 does change it for `incus image copy`, but not for `incus image import`. There it is still done as part of the code of the [cmd in the incus cli](https://github.com/stgraber/incus/blob/main/cmd/incus/image.go#L897-L907) and not in the client.

Reply:
Ah right, that's because `incus image import` doesn't get to send an API struct to the server as we instead send raw image data and rely on headers for the rest.

So having the logic applying the aliases after import makes sense.

We should be able to extend the API to allow for aliases to be passed as HTTP headers too but that still won't be as flexible if we ever want to support alias re-use like we do in the CLI now.

Reply:
https://github.com/lxc/incus/pull/1434

Reply:
@stgraber thanks for the addition in https://github.com/lxc/incus/pull/1434. This will become available with incus 6.8, which is not yet released. But for now, this is not available and I assume, that the incus terraform provider should also be backwards compatible and work with older versions of incus. Therefore I assume, that we stick to the current implementation. Do you agree?
CC: @maveonair

Reply:
Normally I'd agree but given that this is a new feature which won't be in a stable version of the provider for another few weeks, combined with Incus 6.8 being released next week, I think we can just rely on it.

On the Incus side, only the latest release is supported so 6.7 will no longer be officially supported once 6.8 comes out. The exception there is the LTS release, but 6.0.3 will come out a week after 6.8 and will include this API improvement, so we'll have it everywhere we care about.